src/engine/GroupBy.cpp

```diff
@@ -992,8 +969,11 @@ void GroupBy::computeGroupByForHashMapOptimization(
     IdTable* result, std::vector<HashMapAliasInformation>& aggregateAliases,
     size_t numAggregates, const IdTable& subresult, size_t columnIndex,
     LocalVocab* localVocab) {
-  ad_utility::HashMapWithMemoryLimit<KeyType, ValueType> map(
-      getExecutionContext()->getAllocator());
+  // Initialize aggregation data
+  HashMapAggregationData aggregationData{
+      ad_utility::HashMapWithMemoryLimit<KeyType, ValueType>(
+          getExecutionContext()->getAllocator()),
+      std::vector<std::vector<AverageAggregationData>>(numAggregates)};
```

Comment:
Your data class should have private data members, and then a constructor that takes an allocator and the number of aggregates as arguments, then this call gets shorter and safer.
You then also should check inside that constructor, that the `numAggregates >0` because otherwise you run into trouble with your `.at(0)` calls.